autogpt_platform/backend/backend/blocks/smart_decision_maker.py

```diff
@@ -0,0 +1,197 @@
+            tool_sink_nodes = {link.sink_id for link in links}
+            tool_function = {"name": tool_name}
+
+            if len(tool_sink_nodes) == 1:
+                tool_block = backend.blocks.AVAILABLE_BLOCKS[
+                    node_block_map[next(iter(tool_sink_nodes))]
+                ]
+                tool_function["name"] = tool_block().name
+                tool_function["description"] = tool_block().description
+            elif len(tool_sink_nodes) > 1:
+                tool_blocks = [
+                    backend.blocks.AVAILABLE_BLOCKS[node_block_map[node_id]]
+                    for node_id in tool_sink_nodes
+                ]
+                tool_function["name"] = tool_blocks[0]().name
+                tool_function["description"] = (
+                    "This tool is a multi-step tool that can be used to perform a task. "
+                    "It includes blocks that do: "
+                    + ", ".join(
+                        block.name
+                        for block in tool_blocks
+                        if isinstance(block.name, str)
+                    )
+                )
+            else:
+                raise ValueError(
+                    f"Expected at least one tool link in the graph: {tool_links}"
+                )
+
+            properties = {}
+            required = []
+
+            for link in links:
+                required.append(link.sink_name)
+                sink_block = backend.blocks.AVAILABLE_BLOCKS[
+                    node_block_map[link.sink_id]
+                ]
+                sink_block_input = sink_block().input_schema.__fields__[link.sink_name]
+                sink_type = sink_block_input.annotation
+                sink_description = sink_block_input.description
+
+                if sink_type not in ["string", "number", "boolean", "object"]:
+                    sink_type = "string"
+
+                properties[link.sink_name] = {
+                    "type": sink_type,
+                    "description": sink_description,
+                }
+
+            tool_function["parameters"] = {
+                "type": "object",
+                "properties": properties,
+                "required": required,
+                "additionalProperties": False,
+                "strict": True,
+            }
+
+            tool_functions.append({"type": "function", "function": tool_function})
+
+        return tool_functions
+
+    def run(
+        self,
+        input_data: Input,
+        *,
+        graph_id: str,
+        node_id: str,
+        graph_exec_id: str,
+        node_exec_id: str,
+        user_id: str,
+        **kwargs,
+    ) -> BlockOutput:
+        db_client = get_database_manager_client()
+
+        # Retrieve the current graph and node details
+        graph = db_client.get_graph(graph_id)
+
+        if not graph:
+            raise ValueError(f"Graph not found {graph_id}")
+
+        tool_functions = self._create_function_signature(node_id, graph)
```

Comment:
Test this works with mem0 https://github.com/Significant-Gravitas/AutoGPT/blob/016ec0ff6b38341873735a11d2585a8ad8c623e8/autogpt_platform/backend/backend/blocks/mem0.py#L121 since it uses the kwargs based on its place in the graph

Reply:
@ntindle It seems that high level decision is that tools must be an AgentExecutionBlock. 

If you feel strongly that this is a bad idea, let me know and we can discuss it with Toran

Reply:
Why should it require that? I can see plenty of use for this not needing a whole agent and the overhead that would come with it.